votor/src/event_handler.rs

```diff
@@ -387,6 +435,73 @@ impl EventHandler {
         Ok(votes)
     }
 
+    /// Under normal cases we should have a parent ready for first slot of every window.
+    /// But it could be we joined when the later slots of the window are finalized, then
+    /// we never saw the parent ready for the first slot and haven't voted for first slot
+    /// so we can't keep processing rest of the window. This is especially a problem for
+    /// cluster standstill.
+    /// For example:
+    ///    A 40%
+    ///    B 40%
+    ///    C 30%
+    /// A and B finalize block together up to slot 9, now A exited and C joined.
+    /// C sees block 9 as finalized, but it never had parent ready triggered for slot 8.
+    /// C can't vote for any slot in the window because there is no parent ready for slot 8.
+    /// While B is stuck because it is waiting for >60% of the votes to finalize slot 9.
+    /// The cluster will get stuck.
+    /// After we add the following function, C will see that block 9 is finalized yet
+    /// it never had parent ready for slot 9, so it will trigger parent ready for slot 9,
+    /// this means C will immediately vote Notarize for  slot 9, then vote Notarize for
+    /// all later slots. So B and C together can keep finalizing the blocks and unstuck the
+    /// cluster. If we get a finalization cert for later slots of the window and we have the
+    /// block replayed, trace back to the first slot of the window and emit parent ready.
+    fn add_missing_parent_ready(
+        block: Block,
+        ctx: &SharedContext,
+        vctx: &mut VotingContext,
+        local_context: &mut LocalContext,
+    ) -> Option<(Slot, Block)> {
+        let (slot, block_id) = block;
+        let first_slot_of_window = first_of_consecutive_leader_slots(slot);
+        if first_slot_of_window == slot || first_slot_of_window == 0 {
+            // No need to trigger parent ready for the first slot of the window
+            return None;
+        }
+        if vctx.vote_history.highest_parent_ready_slot() >= Some(first_slot_of_window)
+            || !local_context.finalized_blocks.contains(&block)
+        {
+            return None;
+        }
+        // If the block is missing, we can't trigger parent ready
+        let bank_forks_r = ctx.bank_forks.read().unwrap();
+        let bank = bank_forks_r.get(slot)?;
```

Comment:
nit: get already clones the bank so you can just do 
`let bank = ctx.bank_forks.read().unwrap().get(slot)?;`

Reply:
Done.